cdap-data-fabric/src/main/java/co/cask/cdap/data2/security/RemoteUGIProvider.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.net.URL;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Makes requests to ImpersonationHandler to request credentials.
+ */
+public class RemoteUGIProvider implements UGIProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteUGIProvider.class);
+  private static final Gson GSON = new Gson();
+
+  private final Supplier<EndpointStrategy> endpointStrategySupplier;
+  private final LocationFactory locationFactory;
+  private final HttpRequestConfig httpRequestConfig;
+
+  @Inject
+  public RemoteUGIProvider(CConfiguration cConf, final DiscoveryServiceClient discoveryClient,
+                           LocationFactory locationFactory) {
+    this.endpointStrategySupplier = Suppliers.memoize(new Supplier<EndpointStrategy>() {
+      @Override
+      public EndpointStrategy get() {
+        return new RandomEndpointStrategy(discoveryClient.discover(Constants.Service.APP_FABRIC_HTTP));
+      }
+    });
+    this.locationFactory = locationFactory;
+
+    int httpClientTimeoutMs = cConf.getInt(Constants.HTTP_CLIENT_TIMEOUT_MS);
+    this.httpRequestConfig = new HttpRequestConfig(httpClientTimeoutMs, httpClientTimeoutMs);
+  }
+
+  private String resolve(String resource) {
+    Discoverable discoverable = endpointStrategySupplier.get().pick(3L, TimeUnit.SECONDS);
+    if (discoverable == null) {
+      throw new RuntimeException(
+        String.format("Cannot discover service %s", Constants.Service.APP_FABRIC_HTTP));
+    }
+    InetSocketAddress addr = discoverable.getSocketAddress();
+
+    return String.format("http://%s:%s%s/%s",
+                         addr.getHostName(), addr.getPort(), "/v1", resource);
+  }
+
+  private HttpResponse executeRequest(ImpersonationInfo impersonationInfo) {
+    String resolvedUrl = resolve("/impersonation/credentials");
+    try {
+      URL url = new URL(resolvedUrl);
+      HttpRequest.Builder builder = HttpRequest.post(url).withBody(GSON.toJson(impersonationInfo));
+      HttpResponse response = HttpRequests.execute(builder.build(), httpRequestConfig);
+      if (response.getResponseCode() == HttpURLConnection.HTTP_OK) {
+        return response;
+      }
+      throw new RuntimeException(String.format("%s Response: %s.",
+                                               createErrorMessage(resolvedUrl), response));
+    } catch (IOException e) {
+      throw new RuntimeException(createErrorMessage(resolvedUrl), e);
+    }
+  }
+
+  // creates error message, encoding details about the request
+  private static String createErrorMessage(String resolvedUrl) {
```

Comment:
it centralizes the string formatting (used in 2 places). Makes future changes to it consistent across both places.